{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import glob\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>level_0</th>\n",
       "      <th>level_1</th>\n",
       "      <th>level_2</th>\n",
       "      <th>level_3</th>\n",
       "      <th>level_4</th>\n",
       "      <th>level_5</th>\n",
       "      <th>level_6</th>\n",
       "      <th>level_7</th>\n",
       "      <th>##GTF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>3631</td>\n",
       "      <td>5899</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"AT1G01010\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3631</td>\n",
       "      <td>5899</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3760</td>\n",
       "      <td>3913</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3996</td>\n",
       "      <td>4276</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>CDS</td>\n",
       "      <td>4486</td>\n",
       "      <td>4605</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>927385</td>\n",
       "      <td>ChrM</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>exon</td>\n",
       "      <td>230304</td>\n",
       "      <td>230362</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"ATMG01275\"; gene_id \"ATMG01275\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>927386</td>\n",
       "      <td>ChrM</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>exon</td>\n",
       "      <td>233874</td>\n",
       "      <td>234132</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"ATMG01275\"; gene_id \"ATMG01275\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>927387</td>\n",
       "      <td>ChrM</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>gene</td>\n",
       "      <td>264467</td>\n",
       "      <td>265516</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"ATMG01170-2\"; gene_id \"ATMG0117...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>927388</td>\n",
       "      <td>ChrM</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>264467</td>\n",
       "      <td>265516</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"ATMG01170\"; gene_id \"ATMG01170\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>927389</td>\n",
       "      <td>ChrM</td>\n",
       "      <td>Araport11</td>\n",
       "      <td>CDS</td>\n",
       "      <td>264467</td>\n",
       "      <td>265516</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>transcript_id \"ATMG01170\"; gene_id \"ATMG01170\";</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>927390 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       level_0    level_1 level_2  level_3  level_4 level_5 level_6 level_7  \\\n",
       "0         Chr1  Araport11    gene     3631     5899       .       +       .   \n",
       "1         Chr1  Araport11    mRNA     3631     5899       .       +       .   \n",
       "2         Chr1  Araport11     CDS     3760     3913       .       +       0   \n",
       "3         Chr1  Araport11     CDS     3996     4276       .       +       2   \n",
       "4         Chr1  Araport11     CDS     4486     4605       .       +       0   \n",
       "...        ...        ...     ...      ...      ...     ...     ...     ...   \n",
       "927385    ChrM  Araport11    exon   230304   230362       .       +       .   \n",
       "927386    ChrM  Araport11    exon   233874   234132       .       +       .   \n",
       "927387    ChrM  Araport11    gene   264467   265516       .       -       .   \n",
       "927388    ChrM  Araport11    mRNA   264467   265516       .       -       .   \n",
       "927389    ChrM  Araport11     CDS   264467   265516       .       -       0   \n",
       "\n",
       "                                                    ##GTF  \n",
       "0         transcript_id \"AT1G01010\"; gene_id \"AT1G01010\";  \n",
       "1       transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";  \n",
       "2       transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";  \n",
       "3       transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";  \n",
       "4       transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";  \n",
       "...                                                   ...  \n",
       "927385    transcript_id \"ATMG01275\"; gene_id \"ATMG01275\";  \n",
       "927386    transcript_id \"ATMG01275\"; gene_id \"ATMG01275\";  \n",
       "927387  transcript_id \"ATMG01170-2\"; gene_id \"ATMG0117...  \n",
       "927388    transcript_id \"ATMG01170\"; gene_id \"ATMG01170\";  \n",
       "927389    transcript_id \"ATMG01170\"; gene_id \"ATMG01170\";  \n",
       "\n",
       "[927390 rows x 9 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf='/global/scratch/users/chandlersutherland/Athaliana/GTFs/Araport11_GTF_genes_transposons.current.gtf'\n",
    "gtf_import=pd.read_csv(gtf,sep='\\t', skiprows=3).reset_index()\n",
    "gtf_import"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_import=gtf_import.rename(columns={\"level_0\": \"chrom\", \"level_1\": \"Genome\", \"level_2\" : \"type\", \"level_3\" : \"chromStart\", \"level_4\":\"chromEnd\", \"level_6\":'strand'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_import=gtf_import[gtf_import['type'] == 'gene']\n",
    "gene_import['name'] = gene_import['##GTF'].str.split(';',expand=True)[1].str.split(' ', expand=True)[2].str.replace('\"', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_ready = gtf_import.drop(['Genome', 'type', 'level_5', 'level_7', '##GTF'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>3631</td>\n",
       "      <td>5899</td>\n",
       "      <td>+</td>\n",
       "      <td>AT1G01010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>17</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>6788</td>\n",
       "      <td>9130</td>\n",
       "      <td>-</td>\n",
       "      <td>AT1G01020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>138</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>11649</td>\n",
       "      <td>13714</td>\n",
       "      <td>-</td>\n",
       "      <td>AT1G01030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>157</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>23121</td>\n",
       "      <td>31227</td>\n",
       "      <td>+</td>\n",
       "      <td>AT1G01040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>246</td>\n",
       "      <td>Chr1</td>\n",
       "      <td>28500</td>\n",
       "      <td>28706</td>\n",
       "      <td>+</td>\n",
       "      <td>AT1G01046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>927371</td>\n",
       "      <td>ChrM</td>\n",
       "      <td>126289</td>\n",
       "      <td>128223</td>\n",
       "      <td>-</td>\n",
       "      <td>ATMG01390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>927374</td>\n",
       "      <td>ChrM</td>\n",
       "      <td>361637</td>\n",
       "      <td>361939</td>\n",
       "      <td>+</td>\n",
       "      <td>ATMG00650-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>927377</td>\n",
       "      <td>ChrM</td>\n",
       "      <td>82935</td>\n",
       "      <td>83321</td>\n",
       "      <td>-</td>\n",
       "      <td>ATMG01275-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>927381</td>\n",
       "      <td>ChrM</td>\n",
       "      <td>230304</td>\n",
       "      <td>234132</td>\n",
       "      <td>+</td>\n",
       "      <td>ATMG01275-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>927387</td>\n",
       "      <td>ChrM</td>\n",
       "      <td>264467</td>\n",
       "      <td>265516</td>\n",
       "      <td>-</td>\n",
       "      <td>ATMG01170-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33247 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom  chromStart  chromEnd strand         name\n",
       "0       Chr1        3631      5899      +    AT1G01010\n",
       "17      Chr1        6788      9130      -    AT1G01020\n",
       "138     Chr1       11649     13714      -    AT1G01030\n",
       "157     Chr1       23121     31227      +    AT1G01040\n",
       "246     Chr1       28500     28706      +    AT1G01046\n",
       "...      ...         ...       ...    ...          ...\n",
       "927371  ChrM      126289    128223      -    ATMG01390\n",
       "927374  ChrM      361637    361939      +  ATMG00650-2\n",
       "927377  ChrM       82935     83321      -  ATMG01275-2\n",
       "927381  ChrM      230304    234132      +  ATMG01275-3\n",
       "927387  ChrM      264467    265516      -  ATMG01170-2\n",
       "\n",
       "[33247 rows x 5 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_ready"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_ready.to_csv(\"/global/scratch/users/chandlersutherland/Athaliana/Atha_genes.bed\", sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make a CDS only bed file with chromosomes coded as just numbers to play nice with egglib vcf parser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/global/software/sl-7.x86_64/modules/langs/python/3.7/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n",
      "/global/software/sl-7.x86_64/modules/langs/python/3.7/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>strand</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>3760</td>\n",
       "      <td>3913</td>\n",
       "      <td>+</td>\n",
       "      <td>AT1G01010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3996</td>\n",
       "      <td>4276</td>\n",
       "      <td>+</td>\n",
       "      <td>AT1G01010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>4486</td>\n",
       "      <td>4605</td>\n",
       "      <td>+</td>\n",
       "      <td>AT1G01010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>4706</td>\n",
       "      <td>5095</td>\n",
       "      <td>+</td>\n",
       "      <td>AT1G01010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>5174</td>\n",
       "      <td>5326</td>\n",
       "      <td>+</td>\n",
       "      <td>AT1G01010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>927376</td>\n",
       "      <td>M</td>\n",
       "      <td>361637</td>\n",
       "      <td>361939</td>\n",
       "      <td>+</td>\n",
       "      <td>ATMG00650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>927379</td>\n",
       "      <td>M</td>\n",
       "      <td>82937</td>\n",
       "      <td>83321</td>\n",
       "      <td>-</td>\n",
       "      <td>ATMG01275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>927383</td>\n",
       "      <td>M</td>\n",
       "      <td>230304</td>\n",
       "      <td>230362</td>\n",
       "      <td>+</td>\n",
       "      <td>ATMG01275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>927384</td>\n",
       "      <td>M</td>\n",
       "      <td>233874</td>\n",
       "      <td>234132</td>\n",
       "      <td>+</td>\n",
       "      <td>ATMG01275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>927389</td>\n",
       "      <td>M</td>\n",
       "      <td>264467</td>\n",
       "      <td>265516</td>\n",
       "      <td>-</td>\n",
       "      <td>ATMG01170</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>286261 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom  chromStart  chromEnd strand       name\n",
       "2          1        3760      3913      +  AT1G01010\n",
       "3          1        3996      4276      +  AT1G01010\n",
       "4          1        4486      4605      +  AT1G01010\n",
       "5          1        4706      5095      +  AT1G01010\n",
       "6          1        5174      5326      +  AT1G01010\n",
       "...      ...         ...       ...    ...        ...\n",
       "927376     M      361637    361939      +  ATMG00650\n",
       "927379     M       82937     83321      -  ATMG01275\n",
       "927383     M      230304    230362      +  ATMG01275\n",
       "927384     M      233874    234132      +  ATMG01275\n",
       "927389     M      264467    265516      -  ATMG01170\n",
       "\n",
       "[286261 rows x 5 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_import=gtf_import[gtf_import['type'] == 'CDS']\n",
    "cds_import['name'] = cds_import['##GTF'].str.split(';',expand=True)[1].str.split(' ', expand=True)[2].str.replace('\"', '')\n",
    "cds_import['chrom']=cds_import['chrom'].str.replace('Chr', '')\n",
    "\n",
    "bed_ready2 = cds_import.drop(['Genome', 'type', 'level_5', 'level_7', '##GTF'], axis=1)\n",
    "bed_ready2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "bed_ready2.to_csv(\"/global/scratch/users/chandlersutherland/Athaliana/Atha_cds.bed\", sep='\\t', header=False, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
